src/style-spec/expression/definitions/number_format.js

```diff
@@ -0,0 +1,114 @@
+// @flow
+
+import { StringType, NumberType } from '../types';
+
+import type { Expression } from '../expression';
+import type EvaluationContext from '../evaluation_context';
+import type ParsingContext from '../parsing_context';
+import type { Type } from '../types';
+
+declare var Intl: {
+    NumberFormat: Class<Intl$NumberFormat>
+};
+
+declare class Intl$NumberFormat {
+    constructor (
+        locales?: string | string[],
+        options?: NumberFormatOptions
+    ): Intl$NumberFormat;
+
+    static (
+        locales?: string | string[],
+        options?: NumberFormatOptions
+    ): Intl$NumberFormat;
+
+    format(a: number): string;
+
+    resolvedOptions(): any;
+}
+
+type NumberFormatOptions = {
+    style?: 'decimal' | 'currency' | 'percent';
+    currency?: null | string;
+};
+
+export default class NumberFormat implements Expression {
+    type: Type;
+    number: Expression;
+    style: Expression;
+    locale: Expression | null;   // BCP 47 language tag
+    currency: Expression | null; // ISO 4217 currency code, required if style=currency
```

Comment:
Here’s a correspondence to other implementation languages:

Style specification | Objective-C
----|----
`NumberFormat.style` | `NSNumberFormatter.numberStyle`
`NumberFormat.locale` | `NSNumberFormatter.locale`
`NumberFormat.currency` | `NSNumberFormatter.currencyCode`

In Java, use `NumberFormat.get{Currency,Number,Percent}Instance()` and `NumberFormat.setCurrency()`.